// @textlint/config-loader/module/src/loader.js
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
import { isPresetRuleKey } from "./config-util";
import { moduleInterop } from "@textlint/module-interop";
import { loadPreset } from "./preset-loader";
import { isTextlintFilterRuleReporter, isTextlintRuleModule } from "./is";
import { dynamicImport } from "./import";
const isPluginCreator = (mod) => {
    return typeof mod === "object" && Object.prototype.hasOwnProperty.call(mod, "Processor");
};
export const loadPlugins = ({ pluginsObject, moduleResolver, testReplaceDefinitions }) => __awaiter(void 0, void 0, void 0, function* () {
    const plugins = [];
    const pluginErrors = [];
    if (Array.isArray(pluginsObject)) {
        // { plugins: ["a", "b"] }
        yield Promise.all(pluginsObject.map((pluginId) => __awaiter(void 0, void 0, void 0, function* () {
            const resolvedModule = moduleResolver.resolvePluginPackageName(pluginId);
            const mod = yield dynamicImport(resolvedModule.filePath);
            const plugin = moduleInterop(mod.default);
            if (!isPluginCreator(plugin)) {
                pluginErrors.push(new Error(`Plugin should be object that has "Processor" property. But "${pluginId}" is not.

Please check "${pluginId}" is valid plugin.
FilePath: ${resolvedModule.filePath}

For more details, See FAQ: https://github.com/textlint/textlint/blob/master/docs/faq/failed-to-load-textlints-module.md
`));
            }
            plugins.push({
                pluginId,
                plugin,
                filePath: resolvedModule.filePath,
                moduleName: resolvedModule.moduleName
            });
        })));
    }
    else {
        // { plugins: { "a": true, "b": options } }
        yield Promise.all(Object.entries(pluginsObject).map(([pluginId, pluginOptions]) => __awaiter(void 0, void 0, void 0, function* () {
            try {
                // Test Replace logic
                const replacedDefinition = testReplaceDefinitions &&
                    testReplaceDefinitions.find((definition) => {
                        return definition.pluginId === pluginId;
                    });
                if (replacedDefinition) {
                    // for debug
                    plugins.push(replacedDefinition);
                }
                else {
                    const resolvedPlugin = moduleResolver.resolvePluginPackageName(pluginId);
                    const mod = yield dynamicImport(resolvedPlugin.filePath);
                    const plugin = moduleInterop(mod.default);
                    if (!isPluginCreator(plugin)) {
                        pluginErrors.push(new Error(`Plugin should be object that has "Processor" property. But "${pluginId}" is not.

Please check "${pluginId}" is valid plugin.
FilePath: ${resolvedPlugin.filePath}

For more details, See FAQ: https://github.com/textlint/textlint/blob/master/docs/faq/failed-to-load-textlints-module.md
`));
                        return;
                    }
                    plugins.push({
                        pluginId,
                        plugin,
                        options: pluginOptions,
                        filePath: resolvedPlugin.filePath,
                        moduleName: resolvedPlugin.moduleName
                    });
                }
            }
            catch (error) {
                pluginErrors.push(error);
            }
        })));
    }
    return {
        plugins,
        pluginsError: pluginErrors.length === 0
            ? null
            : {
                message: "Can not load plugin",
                errors: pluginErrors
            }
    };
});
export const loadFilterRules = ({ rulesObject, moduleResolver, testReplaceDefinitions }) => __awaiter(void 0, void 0, void 0, function* () {
    // rules
    const rules = [];
    const ruleErrors = [];
    yield Promise.all(Object.entries(rulesObject).map(([ruleId, ruleOptions]) => __awaiter(void 0, void 0, void 0, function* () {
        try {
            // Test Replace logic
            const replacedDefinition = testReplaceDefinitions &&
                testReplaceDefinitions.find((definition) => {
                    return definition.ruleId === ruleId;
                });
            if (replacedDefinition) {
                // for debug
                rules.push(replacedDefinition);
            }
            else {
                const resolvePackage = moduleResolver.resolveFilterRulePackageName(ruleId);
                const mod = yield dynamicImport(resolvePackage.filePath);
                const ruleModule = moduleInterop(mod.default);
                if (!isTextlintFilterRuleReporter(ruleModule)) {
                    ruleErrors.push(new Error(`Filter rule should be object that has "filter" property. But ${ruleId} is not.`));
                    return;
                }
                // rule
                rules.push({
                    ruleId,
                    rule: ruleModule,
                    options: ruleOptions,
                    filePath: resolvePackage.filePath,
                    moduleName: resolvePackage.moduleName
                });
            }
        }
        catch (error) {
            ruleErrors.push(error);
        }
    })));
    return {
        filterRules: rules,
        filterRulesError: ruleErrors.length === 0
            ? null
            : {
                message: "Can not load filter rule",
                errors: ruleErrors
            }
    };
});
export const loadRules = ({ rulesObject, moduleResolver, testReplaceDefinitions }) => __awaiter(void 0, void 0, void 0, function* () {
    // rules
    const rules = [];
    const ruleErrors = [];
    yield Promise.all(Object.entries(rulesObject).map(([ruleId, ruleOptions]) => __awaiter(void 0, void 0, void 0, function* () {
        try {
            // Test Replace logic
            const replacedDefinition = testReplaceDefinitions &&
                testReplaceDefinitions.find((definition) => {
                    return definition.ruleId === ruleId;
                });
            // if rule is disabled, skip to load
            if (!ruleOptions) {
                return;
            }
            if (replacedDefinition) {
                // for debug
                rules.push(replacedDefinition);
            }
            else {
                if (isPresetRuleKey(ruleId)) {
                    // load preset
                    const presetRulesOptions = typeof ruleOptions === "boolean" ? {} : ruleOptions;
                    const rulesInPreset = yield loadPreset({
                        presetName: ruleId,
                        presetRulesOptions,
                        moduleResolver
                    });
                    rules.push(...rulesInPreset);
                }
                else {
                    // load rule
                    const resolvePackage = moduleResolver.resolveRulePackageName(ruleId);
                    const mod = yield dynamicImport(resolvePackage.filePath);
                    const ruleModule = moduleInterop(mod.default);
                    if (!isTextlintRuleModule(ruleModule)) {
                        ruleErrors.push(new Error(`Rule should have "rules" and "rulesConfig" properties. But ${ruleId} is not.
                        
Please check ${ruleId} is valid rule.
FilePath: ${resolvePackage.filePath}

For more details, See FAQ: https://github.com/textlint/textlint/blob/master/docs/faq/failed-to-load-textlints-module.md                        
`));
                        return;
                    }
                    // rule
                    rules.push({
                        ruleId,
                        rule: ruleModule,
                        options: ruleOptions,
                        filePath: resolvePackage.filePath,
                        moduleName: resolvePackage.moduleName
                    });
                }
            }
        }
        catch (error) {
            ruleErrors.push(error);
        }
    })));
    return {
        rules,
        rulesError: ruleErrors.length === 0
            ? null
            : {
                message: "Can not load rule",
                errors: ruleErrors
            }
    };
});
