// @textlint/config-loader/module/src/preset-loader.js
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
// LICENSE : MIT
import { moduleInterop } from "@textlint/module-interop";
import { normalizeTextlintPresetSubRuleKey } from "@textlint/utils";
import { isPresetCreator } from "./is";
import { dynamicImport } from "./import";
export function loadPreset({ presetName, presetRulesOptions, moduleResolver }) {
    return __awaiter(this, void 0, void 0, function* () {
        const presetPackageName = moduleResolver.resolvePresetPackageName(presetName);
        const mod = yield dynamicImport(presetPackageName.filePath);
        const preset = moduleInterop(mod.default);
        if (!isPresetCreator(preset)) {
            throw new Error(`preset should have rules and rulesConfig: ${presetName}`);
        }
        // we should use preset.rules → some preset use different name actual rule
        return Object.keys(preset.rules).map((ruleId) => {
            var _a;
            const normalizedKey = normalizeTextlintPresetSubRuleKey({ preset: presetName, rule: ruleId });
            return {
                ruleId: normalizedKey,
                rule: preset.rules[ruleId],
                options: (_a = presetRulesOptions[ruleId]) !== null && _a !== void 0 ? _a : preset.rulesConfig[ruleId],
                filePath: presetPackageName.filePath,
                moduleName: presetPackageName.moduleName
            };
        });
    });
}
