// textlint/module/src/createLinter.js
// LICENSE : MIT
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
import { TextlintKernel } from "@textlint/kernel";
import { findFiles, pathsToGlobPatterns, separateByAvailability } from "./util/find-util";
import { ExecuteFileBackerManager } from "./engine/execute-file-backer-manager";
import { CacheBacker } from "./engine/execute-file-backers/cache-backer";
import path from "path";
import crypto from "crypto";
import pkgConf from "read-pkg-up";
import fs from "fs/promises";
import { Logger } from "./util/logger";
import debug0 from "debug";
const debug = debug0("textlint:createTextlint");
const createHashForDescriptor = (descriptor) => {
    try {
        const version = pkgConf.sync({ cwd: __dirname }).pkg.version;
        const toString = JSON.stringify(descriptor.toJSON());
        const md5 = crypto.createHash("md5");
        return md5.update(`${version}-${toString}`, "utf8").digest("hex");
    }
    catch (error) {
        // Fallback for some env
        // https://github.com/textlint/textlint/issues/597
        Logger.warn("Use random value as hash because calculating hash value throw error", error);
        return crypto.randomBytes(20).toString("hex");
    }
};
export const createLinter = (options) => {
    var _a, _b;
    const executeFileBackerManger = new ExecuteFileBackerManager();
    const cacheBaker = new CacheBacker({
        cache: (_a = options.cache) !== null && _a !== void 0 ? _a : false,
        cacheLocation: (_b = options.cacheLocation) !== null && _b !== void 0 ? _b : path.resolve(process.cwd(), ".textlintcache"),
        hash: createHashForDescriptor(options.descriptor)
    });
    if (options.cache) {
        executeFileBackerManger.add(cacheBaker);
    }
    else {
        cacheBaker.destroyCache();
    }
    const kernel = new TextlintKernel({
        quiet: options.quiet
    });
    const baseOptions = options.descriptor.toKernelOptions();
    return {
        /**
         * Executes the current configuration on an array of file and directory names.
         * @param {String[]}  files An array of file and directory names.
         * @returns {Promise<TextlintResult[]>} The results for all files that were linted.
         */
        lintFiles(files) {
            return __awaiter(this, void 0, void 0, function* () {
                const patterns = pathsToGlobPatterns(files, {
                    extensions: options.descriptor.availableExtensions
                });
                const targetFiles = findFiles(patterns, {
                    ignoreFilePath: options.ignoreFile
                });
                const { availableFiles, unAvailableFiles } = separateByAvailability(targetFiles, {
                    extensions: options.descriptor.availableExtensions
                });
                debug("Process files", availableFiles);
                debug("No Process files that are un-support extensions:", unAvailableFiles);
                return executeFileBackerManger.process(availableFiles, (filePath) => __awaiter(this, void 0, void 0, function* () {
                    const absoluteFilePath = path.resolve(process.cwd(), filePath);
                    const fileContent = yield fs.readFile(filePath, "utf-8");
                    const kernelOptions = Object.assign({ ext: path.extname(filePath), filePath: absoluteFilePath }, baseOptions);
                    return kernel.lintText(fileContent, kernelOptions);
                }));
            });
        },
        lintText(text, filePath) {
            return __awaiter(this, void 0, void 0, function* () {
                const kernelOptions = Object.assign({ ext: path.extname(filePath), filePath }, baseOptions);
                return kernel.lintText(text, kernelOptions);
            });
        },
        // fix files
        fixFiles(files) {
            return __awaiter(this, void 0, void 0, function* () {
                const patterns = pathsToGlobPatterns(files, {
                    extensions: options.descriptor.availableExtensions
                });
                const targetFiles = findFiles(patterns, {
                    ignoreFilePath: options.ignoreFile
                });
                const { availableFiles, unAvailableFiles } = separateByAvailability(targetFiles, {
                    extensions: options.descriptor.availableExtensions
                });
                debug("Process files", availableFiles);
                debug("No Process files that are un-support extensions:", unAvailableFiles);
                return executeFileBackerManger.process(availableFiles, (filePath) => __awaiter(this, void 0, void 0, function* () {
                    const absoluteFilePath = path.resolve(process.cwd(), filePath);
                    const fileContent = yield fs.readFile(filePath, "utf-8");
                    const kernelOptions = Object.assign({ ext: path.extname(filePath), filePath: absoluteFilePath }, baseOptions);
                    return kernel.fixText(fileContent, kernelOptions);
                }));
            });
        },
        fixText(text, filePath) {
            return __awaiter(this, void 0, void 0, function* () {
                const kernelOptions = Object.assign({ ext: path.extname(filePath), filePath }, baseOptions);
                return kernel.fixText(text, kernelOptions);
            });
        }
    };
};
